### ISSUE TITLE
Configuration Dialect

### ISSUE THREAD

**Issue by andychu** [oilshell/oil#951]
Summary: YAML alternative  / COWS: configuration in Oil without Side effects

This is part of #631: Oil Blocks

Rough sketch for what we need:

- `evalblock()` function
  - it can do PushTemp() and PopTemp() ?
- the COWS evaluator uses `shopt --set dynamic_scope` ?
- `shopt --set parse_equals` -- this is NO LONGER a synonym for const
  -  I think it's like a `setvar` without a corresponding `var`  ???  Inside a block.
  - but is there a difference at the top level ???

Example of evalblock

```
proc p(name, &myblock) {
  var d = evalblock(myblock)
  setvar d->name = name
  
  json write :d
  setvar _cows = d     # does this make sense?
}

p myname {
   age = 30
   ratio = 0.1
}
```

**Comment by andychu** [comment]
Use cases

- [ ] Generating Ninja files.  Because you need it for the repo, and it's an easier way of unifying shell and make
- [ ] travis and sourcehut yaml.  There is going to be SOME platform-specific glue, like maybe the Travis `cache:`, although I want most of it in Ninja, so it's incremental

**Comment by andychu** [comment]
Note: you might need functions, like:

```
func f(ages) {
  for (a in ages) {
    p myname {
      age = a
    } 
  }
}

f(1, 2 3)

```

But does this parse ?

**Comment by andychu** [comment]
Idea for `evalblock()` semantics

- only **consts** are exported.  So `x = 42`, but not `var y = 43`.
- location info is also exported somehow?  So if you validate the value, you can point to an error

Does `value_t` need a span_id then?  I think I would find that useful in some cases, like the dynamic const check?  If it's already defined?

**Comment by andychu** [comment]
Instead of `evalblock()`, we have #1025

**Comment by andychu** [comment]
From Zulip, an idea to make blocks syntactic sugar for dicts, and then gracefully allow more validation with procs:

https://oilshell.zulipchat.com/#narrow/stream/121540-oil-discuss/topic/Using.20oil.20to.20represent.20complex.20data

---

However I'd still like to define the valid first words, so you could have

```
eval ('myfile.oil', procs=%(package user))
```

And then 'package' and 'user' could be ways to define dicts, like

```
package cppunit

user bob
```

would turn into

```
{ "type": "package",
  "name": "cppunit"
}

{ "type": "user",
  "name": "bob"
}


```

**Comment by andychu** [comment]
New minimal idea (in NOTES.txt)


- existing `shvar PATH=''` to disable externals -- fix `hash -d` bug
- - `push-procs`
- `parse_file(path Str) Block`
- `eval_block(b Block) Int` 
  - `eval_block()` recursively takes capital `TASK` to  a shell string (`#!/usr/bin/env oil`)
  - and other blocks to namespaces
  - and does `{"type": "package", "name": "arg1"}` transformations

Is that it?

**Comment by andychu** [comment]
Evaluation rules:

- Top level, and most blocks: `EvalBlockToNamespace`, which is `command_t` -> `Dict[Str, Obj]`
- UPPER blocks like TASK: just leave it alone -- type `command_t`
  - and then provide  a `block_to_string()` function

**Comment by andychu** [comment]
More isolation issues:

- redirects must be disabled
- `source` builtin shouldn't be able to open random files!
- The `parse_config()` function too!

Doh so we need to restrict the builtins dictionary ...

Or we can rely on containers to do this?  

It is an argument for always forking a process ... But then how do we get procs from one place to the other?

I think we don't want `push-procs`.  I think we want to have an inherent `source` before  the config file evaluation

`bin/oven --source mydialect.oil -- myconfig.oil`

It's a bit more complicated that way but probably more general and secure ... we don't want a blacklisting approach; whitelisting is better.

Although `push-procs` is whitelisting of procs -- it's not whitelisting of builtins like `source` though

**Comment by andychu** [comment]
Although `Executor` holds both `self.builtins` and `self.procs`, so we could have 

```
push --builtins echo printf json -- procs myproc {
   ...
}
```

**Comment by andychu** [comment]
First cut of `hay` works!  After a huge commit

https://github.com/oilshell/oil/commit/47c972391969bf59d9d421abad81bfdae74019e2

More:

- [ ] parse hierarchical names, and respect them -- `hay define package/user package/resources`
- [ ] improve error messages -- it assumes an external command if it's not defined
  - maybe `shopt --set hay_eval` or something?  The `haynode` builtin can set it
- [ ] both `source` and `parse_hay()` and need to close the file

More:

- [ ] more docs -- put it in a tour of Oil.  Ruby-like blocks are a major feature
- [ ] make it statically typed and translatable.  This is hard and related to #636

-----

After some usage and feedback:

- [ ] Implement `shopt --set sandbox_{redirect,extern,source}` ?    (under group `sandbox:all` ?)
  - internally this is self.fd_state, self.ext_prog, self.job_state, etc.

**Comment by andychu** [comment]
- [ ] inline hay -- refine the 3 contexts, only 2 of which are sandboxed
- [ ] add basic sandboxing against "reasonable" code, but don't make any guarantees within the same process
- [ ] improve error messages?   it assumes an external command if it's not defined
  - could be solved by "outside hay"
- [ ] Create an end-to-end demo, especially of process-based config file evaluation
- [ ] both `source` and `parse_hay()` and need to close the file


More:

- [ ] more docs
  - [ ]  put it in a tour of Oil.  Ruby-like blocks are a major feature
  - [ ] Variable declaration mutation scope -- needs a section on hay evaluation.  It does `ctx_Temp` which is like `FOO=bar`
  - [ ] `doc/hay.md` should have a quick start
- [ ] make it statically typed and translatable.  This is hard and related to #636

After some usage and feedback:

- [ ] Implement `shopt --set sandbox_{redirect,extern,source}` ?    (under group `sandbox:all` ?)
  - internally this is self.fd_state, self.ext_prog, self.job_state, etc.

**Comment by bar-g** [comment]
Had a look into the new doc/hay skeleton, and as you have mentioned DSLs there, and I still had your comment about the common  push / pop issue in mind (https://github.com/oilshell/oil/issues/1059#issuecomment-1151734946), I felt maybe there is some common solution there to find.

Ruby/crystal seems to have two types of blocks for DSLs:

> The difference between using do ... end and { ... } is that do ... end binds to the left-most call, while { ... } binds to the right-most call. The reason for this is to allow creating Domain Specific Languages (DSLs) using do ... end to have them be read as plain English: [...] You wouldn't want the above to be: [...] (https://crystal-lang.org/reference/1.4/syntax_and_semantics/blocks_and_procs.html)

May be some kind of general apply/with/to/do or and? -block might help for DSLs, too?

**Comment by andychu** [comment]
Yeah there is definitely possibility of things like that ... though I want people to use it first, and actually hit the problem, before we add complexity!

I'm pretty happy with the simple first pass, and hopefully we'll get a bunch of feedback on it

This feature is a bit more obscure than others, because I think platforms / maintainers will generally use it, not end users.   i.e. it's for people building something like sourcehut, Github Actions, systemd, podman/Docker, etc.

**Comment by bar-g** [comment]
Sure, a universal/left/right block application idea/rules will likely need some time to ripe anyway.

I had thought hay could be nice even just for exposing simplest config files to users, thanks to the sandboxing, safe to "source", no need to write parsing code? But I didn't grasp the whole usage/prcractical picture just yet.

**Comment by bar-g** [comment]
Something in your latest "Sketches of YSH Features" blog and your intention for ysh to be well-extensible by users, made me remember my comments here. You wrote:

>  [proc] Arguments are bound left to right, with splats like ... picking up the rest, except for the block argument. It's always the last argument, so it's neither positional or named.

This sounds as if it wouldn't allow to pass multiple blocks. 

I'm not sure if that would still allow for exposing/allowing for more generic block usage, i.e. having left and right binding blocks (for better readable DSLs https://github.com/oilshell/oil/issues/951#issuecomment-1152953400), as well as making use of hay-based config in ysh itself (e.g. no fuzz readable redirection syntax https://github.com/oilshell/oil/issues/1059#issuecomment-1137481928), and generic apply {...} to {...} block application or "operators".

With code+data in the blocks (i.e. hay) could that maybe be a useful basis to allow for implementing, e.g. said matrix algebra operators etc., as extensions as ysh libraries instead of compiled code?

**Comment by andychu** [comment]
I think you can only pass on block in the `{}` style, as the last arg

But there's nothing stopping you from passing multiple blocks the expression style, like

```
myproc (block1, block2)
```

It's mainly a syntactic limitation

But yeah when we implement that, we should have some test cases for it
